client/packages/android/app/src/main/java/org/openmsupply/client/CertWebViewClient.java

```diff
@@ -65,59 +69,105 @@ private X509Certificate get_x509(SslCertificate certificate) {
-    public void onReceivedSslError(WebView view, SslErrorHandler handler,
-                                   SslError error) {
-        // TODO store and trust previous cert
-        if (!error.getUrl().contains("localhost")) {
-            handler.proceed();
-            return;
-        }
+    private boolean validateLocalCertificate(SslCertificate targetCert) {
         // If there is a ssl error, check if the request was trying to reach our local trusted
         // remote server. For this:
         // 1) load self signed remote server certificate from local storage
         // 2) validate that the request was trying to reach the server by validating that the
         //    target certificate from the request is matching our known self signed certificate
         Certificate selfSignedCert = this.get_self_signed_cert();
-        SslCertificate targetCert = error.getCertificate();
         if (selfSignedCert == null || targetCert == null) {
-            super.onReceivedSslError(view, handler, error);
-            return;
+            return false;
         }
         Certificate targetX509Cert = get_x509(targetCert);
         if (targetX509Cert == null) {
             Log.e(TAG, "Failed to extract x509 request target certificate");
-            super.onReceivedSslError(view, handler, error);
-            return;
+            return false;
         }
 
         try {
             targetX509Cert.verify(selfSignedCert.getPublicKey());
-            handler.proceed();
-            return;
+            return true;
         } catch (Exception e) {
             Log.e(TAG, "Invalid request target certificate" + e);
         }
 
-        super.onReceivedSslError(view, handler, error);
+        return false;
+    }
+
+    private boolean validateNonLocalCertificate(SslCertificate targetCert, NativeApi.omSupplyServer connectedServer) {
+        // Calculate SSL fingerprint
+        PublicKey publicKey = get_x509(targetCert).getPublicKey();
+        MessageDigest md = null;
+        try {
+           md = MessageDigest.getInstance("SHA-256");
+        } catch (Exception e) {
+            Log.e(TAG, "SHA-256 algorithm is missing" + e);
+            return false;
+        }
+        md.update(publicKey.getEncoded());
+        String fingerprint = Base64.encodeToString(md.digest(), Base64.DEFAULT).trim();
+
+        // Match SSL fingerprint for server stored in app data
+        // Match by hardware id and port
+        String identifier = connectedServer.getHardwareId() + "-" + connectedServer.getPort();
+        String savedCertFingerprint = savedCertFingerprints.getString(identifier, "");
+        // Save if fingerprint was not found for server
+        if(savedCertFingerprint.length() == 0) {
+            SharedPreferences.Editor editor = savedCertFingerprints.edit();
+            editor.putString(identifier, fingerprint);
+            editor.apply();
+            return true;
+        }
+
+        return savedCertFingerprint.equals(fingerprint);
```

Comment:
We still need to validate that the server really has the private key here, i.e. we have to do the same validation as for  `validateLocalCertificate` just that we use the public key that the server sent us instead of loading it from disc.